### PR TITLE
Support XML response #6 and add global response counter #9

### PR DESCRIPTION
**Pull request by msavy**
```
{
  "method": "GET",
  "path": "/foo/bar",
  "args": "baz=alpha",
  "body": "",
  "counter": 5,
  "headers": {
    "HTTP_VERSION": "HTTP/1.1",
    "HTTP_HOST": "localhost:3000",
    "HTTP_CONNECTION": "keep-alive",
    "HTTP_UPGRADE_INSECURE_REQUESTS": "1",
    "HTTP_USER_AGENT": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.87 Safari/537.36",
    "HTTP_ACCEPT": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
    "HTTP_ACCEPT_ENCODING": "gzip, deflate, sdch, br",
    "HTTP_ACCEPT_LANGUAGE": "en-GB,en;q=0.8,en-US;q=0.6,es;q=0.4"
  }
}
```

```
<echoResponse>
  <method>POST</method>
  <path>/a/b/c</path>
  <counter>9</counter>
  <body>foo</body>
  <headers>
    <header>
      <key>VERSION</key>
      <value>HTTP/1.1</value>
    </header>
    <header>
      <key>HOST</key>
      <value>localhost:3000</value>
    </header>
    <header>
      <key>USER_AGENT</key>
      <value>curl/7.49.1</value>
    </header>
    <header>
      <key>ACCEPT</key>
      <value>application/xml</value>
    </header>
  </headers>
  <args>
    <arg>
      <key>a</key>
      <value>2</value>
    </arg>
  </args>
</echoResponse>
```

@EricWittmann See above. Other features will be separate PR(s).

Yes, the XML is a bit verbose. But that's XML for you :-).

### CODE REVIEW

File: echo_api.rb [hunk 1]
```diff
@@ -29,19 +32,46 @@ def get_headers
...
-
-  content_type 'application/json'
-  JSON.pretty_generate(
-    method: request.request_method,
-    path: request.path,
-    args: request.query_string,
-    body: request.body.read,
-    headers: get_headers()
-  )
+  if request.accept?('application/json')
```
Comment (mikz) [inline echo_api.rb]: This could be a breaking change. IMO we have clients that don't send Accept header at all. So the default response type would be xml.

**Comment by mikz** [comment]
Love it :) I'll be happy to merge if it defaults to JSON as changing the default could break some things.

File: echo_api.rb (hunk 1 above)
Comment (msavy) [inline echo_api.rb]: In my testing, if you don't set the accept header sinatra will return `true`  for t his.

**Comment by msavy** [comment]
It does seem to default to JSON (in my testing):

For `request.accept?('application/json')`:

- Browsers put `*/*` as in their Accept value, which returns true for `application/json`
- If you use no accept header it also seems to return true.
- Most browsers prefer XML in their weightings (see example below).

The only problem might be if someone has set the *wrong* Accept header (i.e. have set it to Accept `application/xml` but actually expect and use JSON). More than happy to juggle things around if you feel that's a possibility.

Example browser accept val:

`text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8`

**Comment by mikz** [comment]
@msavy I just enabled review apps on Heroku. So if you rebase on top of master, it should create a deployment.

**Comment by msavy** [comment]
> edit: I think 280b993 does not close #9 because it is not globally unique. If we start two processes with round robin load balancer, two requests will have the same counter.
> 
> Maybe just going with uuid would make more sense? Possible v1 so it can be ordered ?

By globally unique I meant a shared counter amongst the processes on a single sinatra/rack instance (poor choice of words). 

@EricWittmann Do you need the counter to be globally unique amongst load-balanced instances? In that case I can return a UUID if it matters. I was assuming it'd just be running locally for testing or whatever.

**Comment by EricWittmann** [comment]
I think a shared counter across a cluster would certainly be interesting.  For example, that could help with testing rate limiting perhaps.

That said, the primary idea is probably just uniqueness - so a UUID would be fine.  An ordered UUID would be even cooler (bonus points).

I'd probably go with a UUID in some sort of "id" property and then worry about a possible "counter" property in the theoretical future. :)

**Comment by msavy** [comment]
> I'd probably go with a UUID in some sort of "id" property and then worry about a possible "counter" property in the theoretical future. :)

The current implementation doesn’t have a shared/distributed datastore (unless I’m missing something), so random UUID is probably the only option available at the moment.

**Comment by msavy** [comment]
@EricWittmann out of interest, how would a random UUID differ from the sender just setting their own as a header? 

It's super easy to do; I'm just trying to understand its value.

**Comment by EricWittmann** [comment]
I guess it wouldn't really differ that much.  Just easier for the caller I suppose.

**Comment by mikz** [comment]
Thank you!

**Comment by mikz** [comment]
Deployed http://echo-api.herokuapp.com/

**Pull request merged**